Make the interval skew configurable

thomasliasn opened the pull request:
Previously the verify method would only check the current and the previous interval, which causes problems when the user has positive clock skew.

Example:
- Server time is 10:29:00.
- User time is 10:31:00.
- Server will check intervals 09:59:30-10:00:00,10:00:00-10:30:00, and reject the code with only a couple of seconds skew.

This commit makes the default to check both the previous, the current, and the next interval, and also lets the user change this behavior if desired.

corinnekrych commented on src/main/java/org/jboss/aerogear/security/otp/Totp.java:
```diff
@@ -22,17 +22,16 @@
     private final String secret;
     private final Clock clock;
-    private static final int DELAY_WINDOW = 1;
+    private int skewInterval = 1;
```
let's keep it a constant

abstractj commented:
@thomasliasn I would be more than happy to review this change, but before I would really appreciate if you could remove the trailing spaces not relevant to the change and rebase.

Thanks in advance.

Commit:
Make the interval skew configurable

Previously the verify method would only check the current and the previous interval, which causes problems when the user has positive clock skew.

Example:
- Server time is 10:29:00.
- User time is 10:31:00.
- Server will check intervals 09:59:30-10:00:00,10:00:00-10:30:00, and reject the code with only a couple of seconds skew.

This commit makes the default to check both the previous, the current, and the next interval, and also lets the user change this behavior if desired.

thomasliasn commented:
@abstractj, in case you didn't get a notification, I've pushed your suggested change.

abstractj commented:
@thomasliasn I'm not sure if I follow the real issue here, but seems like we're creating a window of attack. I would like to reproduce the real issue, before get it merged.

The project follow strictly [RFC6238](https://tools.ietf.org/html/rfc6238) to make sure that it will be compatible with third-party projects. Quoting them:

```
When an OTP is generated at the
   end of a time-step window, the receiving time most likely falls into
   the next time-step window.  A validation system SHOULD typically set
   a policy for an acceptable OTP transmission delay window for
   validation.  The validation system should compare OTPs not only with
   the receiving timestamp but also the past timestamps that are within
   the transmission delay.  A larger acceptable delay window would
   expose a larger window for attacks.
```

As any specification, it gives a rise to multiple interpretations, maybe I'm missing the real issue here.

edewit commented:
@abstractj the issue is that the client's time could be ahead and not in sync with the server. I agree with you that we don't / can't take into account such situations as it will open up for attacks. If you want to use OTP you must use NTP to sync your clock.

thomasliasn commented:
As already explained in the commit message, and as @edewit correctly points out, the issue is that users will get their codes rejected if the their clock is ahead of the server. As always there will be a trade-off between security and usability. Ideally users devices will be 100% in sync with the server, but obviously this will not always be the case.

From what I have gathered the industry best-practice is to have a time window before and after the current window, as I have implemented:

When setting up Google Authenticator with https://github.com/google/google-authenticator/tree/master/libpam one is informed that the window is 90 seconds:
> By default, tokens are good for 30 seconds and in order to compensate for
> possible time-skew between the client and the server, we allow an extra
> token before and after the current time. If you experience problems with poor
> time synchronization, you can increase the window from its default
> size of 1:30min to about 4min. Do you want to do so (y/n)

Here (http://jacob.jkrall.net/totp/) the spec referred to by @abstractj is interpreted to mean three windows, one before and one after:
> Aside: latency and clock skew
> The Internet can be slow, and clocks might not match exactly, so some leniency is allowed. RFC6238 
> recommends looking an extra time step in either direction, which essentially opens the window from 30 
> seconds to 90 seconds.

Here the "IBM Security Access Manager for Mobile 8.0.0" can configure the number of intervals to check before and after the current one, the default being 2 or 90 seconds total:
http://www-01.ibm.com/support/knowledgecenter/SSELE6_8.0.0/com.ibm.ammob.doc_8.0/task/ConfiguringOneTimePasswordTOTP.html

abstractj commented:
@thomasliasn what you're mentioning can pretty much be achieved using a custom interval https://github.com/aerogear/aerogear-otp-java/blob/master/src/test/java/org/jboss/aerogear/security/otp/TotpTest.java#L88

thomasliasn commented:
@abstractj I don't think that is the right way to go about it. The library should handle the intervals correctly to handle clock skew as I have suggested, and the clock should be configurable for schemes where one wants interval lengths other than the standard 30s.

abstractj commented:
@thomasliasn I'm sorry Thomas, but I'm not following you. If the API already provides a way to configure the interval, I can't see the real issue here.

This API will cover the RFC and let people choose the acceptance interval. Clock synchronization is out of the scope for this project. 

If you tell me that there's no way for people achieve what you're mentioning, just configuring a custom interval, I will be more than happy to understand this specific scenario. Otherwise, I don't see how we could go further with this pull-request.

thomasliasn commented:
I'll expand on my last comment, hopefully that will help clear up the confusion.

The clock interval length in this library is per default 30 seconds. The clients OTP app and the server must use the same interval length, otherwise codes generated by the client app will not match the one generated by the server. 
Say that I set the interval to e.g. 90 seconds. Still I would have the problem that if the users clock has positive skew, the server would reject the code as invalid. It would however improve the issue slightly as the problematic time as the interval changes from one to the next would occur less often.

But, most standard client apps will not let you configure the interval to something other than 30s, e.g. Google Authenticator! So even if changing the interval length would help a bit, it is not be a viable option as it would greatly increase the effort of explaining and guiding the users to make this extra configuration on non-standard apps, and would reduce the adoption rate of two-factor authentication.

What does fix the described skew issue is to check the interval following the current one as I have implemented, and as is the industry standard as shown in my earlier comment.

abstractj commented:
@thomasliasn I'm very sorry, but this to me is pretty much what https://github.com/aerogear/aerogear-otp-java/blob/master/src/main/java/org/jboss/aerogear/security/otp/Totp.java#L102 already does. If you consider this a bug, please, introduce new test case scenarios to reproduce the bug. But please don't change the whole test suite, unless the whole codebase is completely wrong.

Unless you provide the failing test case scenarios which this API can't cover.  I don't see how to move further.

thirunavukarasut commented:
Hi @abstractj, we encountered an issue similar to the one mentioned by @thomasliasn, if the code change is not possible, would it be possible to make the Totp.generate() method as **protected** instead of **_private_**. So that we could override the verify() method with an additional parameter for DELAY_WINDOW, which would help us in resolve the issue.

colinmcdonald22 commented:
@abstractj Recently ran into the issue fixed by this PR. Although it is possible for a custom interval to be used to avoid the problem, that solution becomes very repetitive (say we want to allow an interval larger than +/- 30s) and isn't as intuitive to someone using the API as a `setSkewInterval` method would be.

Thanks.

mryf323 commented:
Hi. I want to use your library at a server back-end. I used your code in two separated function:
1st I generate the code by calling this function:
```
public String sendVerificationCode(int uid) {
        String encoded = Base32.encode(Integer.toString(uid).getBytes());
        System.out.println("encoded:" + encoded);
        Totp totp = new Totp(encoded, new Clock(keySurvivalTime * 60 * 1000));
        String key = totp.now();
        return key;
    }

```
 2nd I call this function to validate the key:
```
public boolean verifyKey(String key, int uid) {
        String encoded = Base32.encode(Integer.toString(uid).getBytes());
        System.out.println("encoded:" + encoded);
        Totp totp = new Totp(encoded, new Clock(keySurvivalTime * 60 * 1000));
        return totp.verify(key);
    }
```
Even after `keySurvivalTime` I get true, I think that's because the clock really counts (mili)seconds in it.
Can you describe my misuse of the code?

danielpassos commented:
This project is not maintained for a long time and going to be archived.
